tests/nosetests/pyanaconda_tests/module_subscription_tests.py

```diff
@@ -0,0 +1,261 @@
+# Red Hat, Inc.
+#
+# Red Hat Author(s): Martin Kolman <mkolman@redhat.com>
+#
+import os
+import unittest
+from unittest.mock import Mock
+import tempfile
+
+from dasbus.typing import *  # pylint: disable=wildcard-import
+
+from pyanaconda.modules.common.constants.services import SUBSCRIPTION
+from pyanaconda.modules.common.structures.subscription import SystemPurposeData
+from pyanaconda.modules.subscription.subscription import SubscriptionModule
+from pyanaconda.modules.subscription.subscription_interface import SubscriptionInterface
+from pyanaconda.modules.subscription.system_purpose import get_valid_fields, normalize_field, \
+    match_field
+from tests.nosetests.pyanaconda_tests import check_kickstart_interface
+
+# content of a valid populated valid values json file for system purpose testing
+SYSPURPOSE_VALID_VALUES_JSON="""
+{
+"role" : ["role_a", "role_b", "role_c"],
+"service_level_agreement" : ["sla_a", "sla_b", "sla_c"],
+"usage" : ["usage_a", "usage_b", "usage_c"]
+}
+"""
+
+# content of a valid but not populated valid values json file for system purpose testing
+SYSPURPOSE_VALID_VALUES_JSON_EMPTY="""
+{
+"role" : [],
+"service_level_agreement" : [],
+"usage" : []
+}
+"""
+
+class SubscriptionInterfaceTestCase(unittest.TestCase):
+    """Test DBus interface for the subscription module."""
+
+    def setUp(self):
+        """Set up the subscription module."""
+        self.subscription_module = SubscriptionModule()
+        self.subscription_interface = SubscriptionInterface(self.subscription_module)
+
+        # Connect to the properties changed signal.
+        self.callback = Mock()
+        self.subscription_interface.PropertiesChanged.connect(self.callback)
+
+    def kickstart_properties_test(self):
+        """Test kickstart properties."""
+        self.assertEqual(self.subscription_interface.KickstartCommands, ["syspurpose", "rhsm"])
+        self.assertEqual(self.subscription_interface.KickstartSections, [])
+        self.assertEqual(self.subscription_interface.KickstartAddons, [])
+        self.callback.assert_not_called()
+
+    def system_purpose_data_test(self):
+        """Test the SystemPurposeData DBus structure."""
+
+        # create the SystemPurposeData structure
+        system_purpose_data = SystemPurposeData()
+        system_purpose_data.role = "foo"
+        system_purpose_data.sla = "bar"
+        system_purpose_data.usage = "baz"
+        system_purpose_data.addons = ["a", "b", "c"]
+
+        # create its expected representation
+        expected_dict = {
+            "role" : get_variant(Str, "foo"),
+            "sla" : get_variant(Str, "bar"),
+            "usage" : get_variant(Str, "baz"),
+            "addons" : get_variant(List[Str], ["a", "b", "c"])
+        }
+
+        # compare the two
+        self.assertDictEqual(SystemPurposeData.to_structure(system_purpose_data), expected_dict)
+
+    def default_property_values_test(self):
```

Comment:
This test is a mess. Please, focus on testing units of the code. There should be always one test dedicated to testing one member of the DBus API of the Subscription module. So one test for the DBus property `SystemPurposeData`, one test the DBus method `GetValidRoles`, etc. Then there should be kickstart tests for the most common use cases. And after all that you should check the coverage of the code and if there are use cases that are not covered, you can add and extend tests for these use cases.

Otherwise, we can end up with a lot of useless tests, but miss the most critical ones.

Reply:
Yeah, this should be already covered by the individual API tests. And another issue is keeping this one test actually up to date when additional methods/properties are added. So let's scrap this. :)